src/ninja.cc

```diff
@@ -662,7 +662,65 @@ void EncodeJSONString(const char *str) {
   }
 }
 
-int NinjaMain::ToolCompilationDatabase(const Options* options, int argc, char* argv[]) {
+enum EvaluateCommandMode {
+  ECM_NORMAL,
+  ECM_EXPAND_RSPFILE
+};
+string EvaluateCommandWithRspfile(Edge* edge, EvaluateCommandMode mode) {
+  string command = edge->EvaluateCommand();
+  if (mode == ECM_NORMAL)
+    return command;
+
+  string rspfile = edge->GetUnescapedRspfile();
+  if (rspfile.empty())
+    return command;
+
+  size_t index = command.find(rspfile);
+  if (index == 0 || index == string::npos || command[index - 1] != '@')
+    return command;
+
+  string rspfile_content = edge->GetBinding("rspfile_content");
+  size_t newline_index = 0;
+  while ((newline_index = rspfile_content.find('\n', newline_index)) !=
```

Comment:
How do you end up with rspfile_content that contains a newline? In the manifest you say

```
   rspfile_content = aasdfsdaf $
                     asdfasdf
```

and ninja replaces the '$\n' with nothing at manifest parse time as far as I remember. How do you get a newline in there?

(This is possibly a silly question; please bear with me.)

Reply:
At the time I couldn't rule out that there wouldn't be newlines, and if there were, then the resulting command line could be malformed. Also, Microsoft's tools allow newlines in response files, so I was hedging the risk of newlines being included.

I can remove the newline handling if the `rspfile_content` lookup is guaranteed to not have any newlines. I haven't looked at the code closely to make that call.

Reply:
Oh, I guess you can get newlines if you do `rspfile_content = $in_newline`. But other than that magic variable, I don't think it's possible. Hrm, annoying, but for that it is possible. (in_newline is only needed in practice for link.exe so it shouldn't appear in compilations either, but…).